### DISCUSSION TITLE
Serverless functions exclude libraries

### DISCUSSION THREAD

**Discussion by Ahmadh26** [vercel/next.js#43629] [Help]
how can i exclude files if a serverless function is exceeding the allowed limit 

here is the error i am getting: 

```bash
Serverless Function's page: api/v1/images/[type]/delete.js
Large Dependencies            Uncompressed size  Compressed size
node_modules/fast-geoip/data             164 MB          28.1 MB
node_modules/.prisma/client             44.4 MB          15.6 MB
public/assets/images                    12.1 MB          11.1 MB
public/assets/fonts                     5.26 MB          2.45 MB
node_modules/next/dist                  4.24 MB          1.28 MB
node_modules/react-dom/cjs              2.28 MB           558 kB
node_modules/@prisma/client             1.29 MB           274 kB

Created all serverless functions in: 38.879s
Collected static files (public/, static/, .next/static): 48.791ms
Build Completed in /vercel/output [5m]
Failed to process build result for "api/v1/images/[type]/delete". Data: {"type":"Lambda"}.
Error: The Serverless Function "api/v1/images/[type]/delete" is 58.57mb which exceeds the maximum size limit of 50mb. Learn More: https://vercel.link/serverless-function-size
```

I have tried creating a vercel.json and doing the following: 

```json
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "functions": {
    "pages/api/**/*.ts": {
      "excludeFiles": "node_modules/fast-geoip/**"
    }
  }
}

```

but that didn't work. how can i resolve this issue..

__note:__ I understand that vercel has geolocation by default but it is not what i want to use at the moment as i self host the website. i am just trying to deploy on vercel to test some stuff.

**Comment by icyJoseph** [comment]
Isn't the issue that `api/v1/[type]/delete` is itself loading up more than 50mb? You need to cut down the amount of data loaded there.

Perhaps you are using something like:

```js
import { something } from 'fast-geoip';
```

Perhaps you could do:

```js
import something from 'fast-geoip/path-to-something'
```

This would happen with some SDK's when importing from the top of the module. These SDKS export JSON and other assets from the root, and importing something from the root unintentionally brings the assets into the module.

**Reply by Ahmadh26** [reply]
The thing is `api/v1/images/[type]/delete` isn't loading fast geoIP at all.. 

here is the API file: 

```ts
import type { NextApiRequest, NextApiResponse } from 'next'
import { promises as fs } from 'fs'
import path from 'path'
import { File } from 'formidable'
import {
  deleteImageSchema,
  ImagesDeleteSchemaParams,
} from 'schema/images.schema'
import { validate } from '@middleware/validateResources'
import { sessionData } from '@interfaces/account'

/* Don't miss that! */
export const config = {
  api: {
    bodyParser: false,
  },
}

type ProcessedFiles = Array<[string, File]>

const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  const { session } = req.body as { session: sessionData }
  const { img, type } = req.query as ImagesDeleteSchemaParams
  const targetPath = path.join(process.cwd(), `public/${img}`)

  try {
    await fs.access(targetPath)
    await fs.rm(targetPath)
    // await fs.mkdir(targetPath, { recursive: true })
  } catch (e) {
    // await fs.mkdir(targetPath, { recursive: true })
    return res.status(400).send({
      message: 'Image not found',
    })
  }

  return res.status(200).send({
    message: 'Image deleted',
  })
}

export default validate(deleteImageSchema, handler, true, ['DELETE'])

```

even the validate function doesn't call fast geoIP in any way or shape as this is what it imports:  (this is the validateResources.tsx file)

```ts
import { Method } from '@interfaces/enums'
import { NextApiHandler, NextApiRequest, NextApiResponse } from 'next'
import { Session, unstable_getServerSession } from 'next-auth'
import { authOptions } from 'pages/api/auth/[...nextauth]'
import { ObjectShape, OptionalObjectSchema } from 'yup/lib/object'
```

**Reply by icyJoseph** [reply]
MMm I'd start by doing `import type { File } from 'formidable'` and likewise everywhere else you only import types.

It is hard to tell from this much code, but importing a type might be also importing heavy assets through several levels of modules.

**Reply by Ahmadh26** [reply]
> MMm I'd start by doing `import type { File } from 'formidable'` and likewise everywhere else you only import types.
> 
> It is hard to tell from this much code, but importing a type might be also importing heavy assets through several levels of modules.

thats understandable, i'll refactor the code to do that.

 i'm more than happy to share more code with you in order to know why is fast-geoIP required in a serverless function that doesnt need it. but the issue is i do not want to share the code publicly here.

**Reply by icyJoseph** [reply]
One thing I forgot, run @next/bundle-analyzer to see what's getting packaged into the server bundles, from there - you should be able to tell which package needs to be trimmed.
